lib/rubocop/cop/rspec/instance_spy.rb

```diff
@@ -0,0 +1,66 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module RSpec
+      # Checks for `instance_double` used with `have_received`.
+      #
+      # @example
+      #   # bad
+      #   it do
+      #     foo = instance_double(Foo).as_null_object
+      #     expect(foo).to have_received(:bar)
+      #   end
+      #
+      #   # good
+      #   it do
+      #     foo = instance_spy(Foo)
+      #     expect(foo).to have_received(:bar)
+      #   end
+      #
+      class InstanceSpy < Cop
+        MSG = 'Use `instance_spy` when you check your double ' \
+              'with `have_received`'.freeze
+
+        EXAMPLES = Examples::ALL.node_pattern_union.freeze
+
+        def_node_matcher :example?, "(block $(send nil #{EXAMPLES}) ...)"
+
+        def_node_search :null_double, <<-PATTERN
+        (lvasgn $_
+          (send
+            $(send nil :instance_double
+              ...) :as_null_object))
+        PATTERN
+
+        def_node_search :have_received_usage, <<-PATTERN
+        (send
+          (send nil :expect
+          (lvar $_)) :to
+          (send nil :have_received
+          ...)
+        ...)
+        PATTERN
+
+        def on_block(node)
+          return unless example?(node)
+
+          null_double(node) do |var, receiver|
+            have_received_usage(node) do |expected|
+              add_offense(receiver, :expression, MSG) if expected == var
+            end
+          end
+        end
+
+        def autocorrect(node)
+          lambda do |corrector|
+            replacement = 'instance_spy'
+            corrector.replace(node.loc.selector, replacement)
+            # '.as_null_object'.length => 15
```

Comment:
In fact, this approach is important too. I know it seems silly but this *is* valid ruby:

```ruby
instance_double(Foo, :name)       .              as_null_object
```